File: BinaryTreeFromPreorderInorder.java
// Time Complexity: O(n)
// Space Complexity: O(n)

import java.util.Arrays;

class BinaryTreeFromPreorderInorder {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        //base
        if(preorder.length == 0) return null;
        //logic
        int rootVal = preorder[0];
        int rootidx = -1;
        TreeNode treeNode = new TreeNode();
        for(int i = 0; i < inorder.length; i++){
            if(inorder[i] == rootVal){
                rootidx = i;
            }
        }

        int[] inleft = Arrays.copyOfRange(inorder, 0, rootidx);
        int[] inright = Arrays.copyOfRange(inorder, rootidx + 1, inorder.length);
        int[] preleft = Arrays.copyOfRange(preorder, 1, inleft.length + 1);
        int[] preright = Arrays.copyOfRange(preorder, inleft.length + 1, preorder.length);

        treeNode.val = rootVal;
        treeNode.left = buildTree(preleft, inleft);
        treeNode.right = buildTree(preright, inright);

        return treeNode;
    }
}

File: binary_tree_from_preorder_inorder.py
'''
Time complexity: O(2n*n)
Space complexity: O(max_depth)
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if len(preorder) == 0:
            return None
        
        root = TreeNode(preorder[0]) # root will always be first element of preorder
        index = -1
        
        # find index of root in inorder
        for x in range(0, len(inorder)):
            if inorder[x] == root.val:
                index = x
                break
        
        preLeft = preorder[1: index + 1] #preoder(1, index+1) => 1 inlcusive, index + 1 exclusive
        preRight = preorder[index + 1: len(preorder)]
        inLeft = inorder[0: index + 1]
        inRight = inorder[index + 1: len(inorder)]
        
        root.left = self.buildTree(preLeft, inLeft)
        root.right = self.buildTree(preRight, inRight)
        
        return root
        